Add fnal booster input files

cemitch99 opened the pull request:
Added MAD-X + SXF input files for a reduced thin-multipole model of the Fermilab Booster lattice for space charge studies in ImpactX.  For this simplified example, space charge is turned off.  A follow-up PR will include the corresponding example with space charge turned on.

Contains a placeholder draft of a script for parsing the SXF input file (testparse.py), to be developed.

- [x] Develop Python script for parsing MAD-X + SXF input file.
- [x] Modify the above Python script to run ImpactX.
- [x] Add parsing of thin dipoles as a special case of multipole m = 0.
- [x] Add "tilt" parameter to Python element specifications where needed.
- [x] Add basic test of execution.
- [x] Double-check all element input definitions.
- [x] Benchmark Twiss functions, tunes, etc.
- [x] Update docs.
- [x] Upload `fermi-booster-madx-sxf_corr_new` (1.4MB text) to Zenodo 
- [ ] Set up CI to download Zenodo data on the fly ([as in ML example](https://github.com/ECP-WarpX/impactx/blob/24.06/examples/pytorch_surrogate_model/run_ml_surrogate_15_stage.py#L90-L101))

Complete data located on Zenodo:  https://doi.org/10.5281/zenodo.11645618

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Add MAD-X, SXF input files for Fermilab Booster.

cemitch99 committed:
Delete input_kurth_10nC.in

This is not part of this PR.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Write SXF Parser

Parse all elements.

Co-authored-by: Chad Mitchell <ChadMitchell@lbl.gov>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add F. Schmidt updated SXF file and new Python script.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Delete examples/fnal_booster/fermi-booster-madx-sxf

Old file, deleted and replaced by file of a similar name.

cemitch99 commented:
There are a large number of MAD-X files here that are not used by ImpactX.  Here, we use only the SXF file.  I am tempted to delete these for clarity, and keep them on the Google Drive for later use by us if needed.

cemitch99 commented on examples/fnal_booster/testparse.py:
```diff
@@ -0,0 +1,77 @@
+
+import re
+
+import pandas as pd
```
Outdated.

cemitch99 committed:
Fix merge.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Delete examples/fnal_booster/main_deb_mad_6c.madx

Delete unused files.

cemitch99 committed:
Delete examples/fnal_booster/match_chrom.madx

Delete unused files.

cemitch99 committed:
Delete examples/fnal_booster/ring_seq_bb_spch_thin.madx

Delete unused files.

cemitch99 committed:
Delete examples/fnal_booster/match_tunes.madx

Delete unused files.

cemitch99 committed:
Delete examples/fnal_booster/spcharge_beam_command.madx

Delete unused files.

cemitch99 committed:
Update fnal_booster.py

Add dipedge rotation tilt to lattice file.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add FNAL Booster execution test.

cemitch99 committed:
Merge branch 'development' into add_fnal_booster

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented:
Need to update paths to the SXF file "fermi-booster-madx-sxf_corr_new" to allow proper access from Python script during CI.

ax3l commented:
You can add something similar to:
```cmake
# copy MAD-X lattice file
file(COPY ${ImpactX_SOURCE_DIR}/examples/fodo/fodo.madx
     DESTINATION ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/FODO.MADX.py)
```

in `examples/CMakeLists.txt` for the SXF file to be available during `ctest` runs.

cemitch99 commented on examples/fnal_booster/fnal_booster.py:
```diff
@@ -0,0 +1,275 @@
+
+                for i in range(len(kl)):
+                    if i == 0:
+                        print(f"elements.ThinDipole(theta={kl[i]}, rc=1.0)")
```
```suggestion
                        angledeg = kl[i] * 180.0 / math.pi
                        print(f"elements.ThinDipole(theta={angledeg}, rc=1.0)")
```

cemitch99 committed:
Update examples/fnal_booster/fnal_booster.py

Convert thin dipole bend angle to degrees.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Include SXF file for ctest runs.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Correct fnal_booster dir name for ctest

cemitch99 committed:
Update CMakeLists.txt

Adjusting supplementary file path for ctest.

cemitch99 committed:
Update CMakeLists.txt

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_fnal_booster

ax3l committed:
Fix: CMake Copy, Imports, Global Var

ax3l committed:
Test: Mark as Slow

cemitch99 committed:
Update fnal_booster.py

Correct ThinDipole bend angle to reflect input degrees.

cemitch99 committed:
Update fnal_booster.py

Correct index of multipole coefficients (+1).

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update fnal_booster.py

Modify ThinDipole and DipEdge parameters to set unknown parameters to reasonable default values.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update fnal_booster.py

Modified initial beam moments to match 1) emittance in MADX initial input line & 2) sigx, sigy from initial space charge element in SXF input.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add new SXF input file and corrected Python parsing script.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Modified filenames for input and SXF files in CMakeLists.txt

cemitch99 committed:
Delete examples/fnal_booster/fnal_booster.py

Deleted obsolete version of input file.

cemitch99 committed:
Update CMakeLists.txt

Correct sxf filename.

cemitch99 committed:
Update fnal_booster_no-sc.py

Update sigma -> lambda

cemitch99 committed:
Update fnal_booster_no-sc.py

Reduced the number of particles to 10,000 for CI tests.

cemitch99 committed:
Add README and analysis.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Link to documentation.
